File: usth/ICT2.2/labwork/3/C++/name-card.cc
#include <regex>
#include <stdexcept>
#include <string>

#include "name-card.h"

using namespace std;

const regex name_pattern ("[ A-Za-z]+");
const regex phone_pattern ("[-0-9]+");
// This should be good enough I guess?
const regex email_pattern ("(\\w+)(\\.|_)?(\\w*)@(\\w+)(\\.(\\w+))+");

void
NameCard::set_name (string s)
{
  // I miss Raku so much
  smatch m;
  if (!regex_match (s, m, name_pattern))
    throw invalid_argument{"invalid name"};
  name = s;
}

void
NameCard::set_phone (string s)
{
  smatch m;
  if (!regex_match (s, m, phone_pattern))
    throw invalid_argument{"invalid number"};
  phone = s;
}

void
NameCard::set_email (string s)
{
  smatch m;
  if (!regex_match (s, m, email_pattern))
    throw invalid_argument{"invalid name"};
  email = s;
}

NameCard::NameCard (string n, string p, string e)
{
  set_name (n);
  set_phone (p);
  set_email (e);
}

File: usth/ICT2.2/labwork/3/Java/NameCard.java
import java.util.regex.Pattern;

public class NameCard
{
  private static final Pattern namePattern = Pattern.compile("[ A-Za-z]+");
  private static final Pattern phonePattern = Pattern.compile("[-0-9]*");
  // I have not learnt regex properly
  private static final Pattern emailPattern = Pattern.compile(
    "(\\w+)(\\.|_)?(\\w*)@(\\w+)(\\.(\\w+))+");

  private String name;
  private String phone;
  private String email;

  public String getName()
  {
    return name;
  }

  public String getPhone()
  {
    return phone;
  }

  public String getEmail()
  {
    return email;
  }

  public void setName(String name)
  {
    if (!namePattern.matcher(name).matches())
      throw new IllegalArgumentException("invalid name: " + name);
    this.name = name;
  }

  public void setPhone(String phone)
  {
    if (!phonePattern.matcher(phone).matches())
      throw new IllegalArgumentException("invalid phone number: " + phone);
    this.phone = phone;
  }

  public void setEmail(String email)
  {
    if (!emailPattern.matcher(email).matches())
      throw new IllegalArgumentException("invalid email: " + email);
    this.email = email;
  }

  public NameCard(String name)
  {
    this(name, "", "");
  }

  public NameCard(String name, String phone)
  {
    this(name, phone, "");
  }

  public NameCard(String name, String phone, String email)
  {
    setName(name);
    setPhone(phone);
    setEmail(email);
  }
}
